src/main/java/io/divolte/server/config/DivolteConfiguration.java

```diff
@@ -1,37 +1,131 @@
 package io.divolte.server.config;
 
 import com.fasterxml.jackson.annotation.JsonCreator;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.*;
+import io.divolte.server.config.constraint.MappingSourceSinkReferencesMustExist;
+import io.divolte.server.config.constraint.OneSchemaPerSink;
+import io.divolte.server.config.constraint.SourceAndSinkNamesCannotCollide;
 
 import javax.annotation.ParametersAreNonnullByDefault;
 import javax.validation.Valid;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 @ParametersAreNonnullByDefault
+@MappingSourceSinkReferencesMustExist
+@SourceAndSinkNamesCannotCollide
+@OneSchemaPerSink
 public final class DivolteConfiguration {
-    @Valid public final ServerConfiguration server;
-    @Valid public final TrackingConfiguration tracking;
-    @Valid public final JavascriptConfiguration javascript;
-    @Valid public final IncomingRequestProcessorConfiguration incomingRequestProcessor;
-    @Valid public final KafkaFlusherConfiguration kafkaFlusher;
-    @Valid public final HdfsFlusherConfiguration hdfsFlusher;
+    @Valid public final GlobalConfiguration global;
+
+    // Mappings, sources and sinks are all keyed by their name.
+    @Valid public final ImmutableMap<String,MappingConfiguration> mappings;
+    @Valid public final ImmutableMap<String,SourceConfiguration> sources;
+    @Valid public final ImmutableMap<String,SinkConfiguration> sinks;
 
     @JsonCreator
-    private DivolteConfiguration(
-            final ServerConfiguration server,
-            final TrackingConfiguration tracking,
-            final JavascriptConfiguration javascript,
-            final IncomingRequestProcessorConfiguration incomingRequestProcessor,
-            final KafkaFlusherConfiguration kafkaFlusher,
-            final HdfsFlusherConfiguration hdfsFlusher) {
-        this.server = server;
-        this.tracking = tracking;
-        this.javascript = javascript;
-        this.incomingRequestProcessor = incomingRequestProcessor;
-        this.kafkaFlusher = kafkaFlusher;
-        this.hdfsFlusher = hdfsFlusher;
+    DivolteConfiguration(final GlobalConfiguration global,
+                         final Optional<ImmutableMap<String, SourceConfiguration>> sources,
+                         final Optional<ImmutableMap<String, SinkConfiguration>> sinks,
+                         final Optional<ImmutableMap<String,MappingConfiguration>> mappings) {
+        this.global = Objects.requireNonNull(global);
+        this.sources = sources.orElseGet(DivolteConfiguration::defaultSourceConfigurations);
+        this.sinks = sinks.orElseGet(DivolteConfiguration::defaultSinkConfigurations);
+        this.mappings = mappings.orElseGet(() -> defaultMappingConfigurations(this.sources.keySet(), this.sinks.keySet()));
+    }
+
+    public BrowserSourceConfiguration getBrowserSourceConfiguration(final String sourceName) {
```

Comment:
I have a preference toward creating a `asBrowserSourceConfiguration()` method on `SourceConfiguration` (and onward `asXxxConfiguration()` for others to come), instead of doing both the presence check and the type check here. This would leave the check for existence of the source (using `Map#containsKey(...)`) to the client code and only performs the type check as part of the config class' logic. In different contexts the calling code might have more meaningful error messages for a missing key compared to the general case.